File: CJacHash/CJacHash/CJacHash/jachash.c
#include <stdio.h>
#include <string.h>
#include "jachash.h"

#define MAX_LENGTH 16

static char transformByte(struct jachash_context *context, char bl);
static char shiftLeft(char b, char bits);
static int pad(char *source, int size);
static void init(struct jachash_context *context);

char *computeHashFromString(struct jachash_context *context, const char* string) {
	char bytes[strlen(string)];
	int i;
	for (i = 0; i < strlen(string); i++)
		bytes[i] = (char)string[i];
	computeHashFromBytes(context, bytes, strlen(string));
}

char *computeHashFromBytes(struct jachash_context *context, char *bytes, int length) {
	init(context);
	int i;
	char result[MAX_LENGTH];
	int sourceSize = pad(bytes, length);
	for (i = 0; i < sourceSize; i++)
		context->x += (char)bytes[i];
	for (i = 0; i < sourceSize; i++)
		result[i % MAX_LENGTH] = transformByte(context, bytes[i]);
	for (i = 0; i < MAX_LENGTH; i++)
		printf("%02x", result[i] & 0xFF);
	printf("\n");
}

static char transformByte(struct jachash_context *context, char bl) {
	context->a = shiftLeft(bl, context->x);
	context->b = (context->b ^ bl) - context->x;
	context->c = (context->a + context->b) & context->x;
	context->d ^= context->x - context->b;
	context->x ^= context->d;
	return ((context->a * context->c) + context->b - context->x * context->d ^ bl);
}

static char shiftLeft(char b, char bits) {
	return ((b << bits | b >> 32 - bits));
}

static int pad(char *source, int size) {
	if (size >= MAX_LENGTH)
		return size;
	int i;
	for (i = size; i < MAX_LENGTH; i++)
		source[i] = 0xFF;
	//source[i] = 0;
	return MAX_LENGTH;
}

static void init(struct jachash_context *context) {
	context->a = 0x6B87;
	context->b = 0x7F43;
	context->c = 0xA4Ad;
	context->d = 0xDC3F;
	context->x = 0;
}

File: src/JJacHash/src/JacHash.java
import java.io.FileInputStream;
import java.io.IOException;


public class JacHash {
	public int MAX_LENGTH = 16;
	public byte FILLER_BYTE = 0xF;
	
	private int a;
	private int b;
	private int c;
	private int d;
	private int x;
	
	public JacHash() {
	}
	
	public String computeHashFromString(String text) {
		return computeHashFromBytes(text.getBytes());
	}
	
	public String computeHashFromBytes(byte[] bytes) {
		init();
		byte[] source = pad(bytes);
		byte[] result = new byte[MAX_LENGTH];
		for (int i = 0; i < source.length; i++)
			x += source[i];
		for (int i = 0; i < source.length; i++)
			result[i % MAX_LENGTH] = transformByte(source[i]);
		String accum = "";
		for (int i = 0; i < MAX_LENGTH; i++)
			accum += String.format("%02x", result[i]);
		return accum;
	}
	
	public String computeHashFromFile(String filePath) throws IOException {
		init();
		FileInputStream reader = new FileInputStream(filePath);
		long length = reader.getChannel().size();
		int appendToStream = 0;
		if (length < MAX_LENGTH)
			appendToStream = MAX_LENGTH;
		while (reader.getChannel().position() < length)
			x += reader.read();
		reader.getChannel().position(0);
		for (int i = 0; i < appendToStream; i++)
			x += FILLER_BYTE;
		byte[] result = new byte[MAX_LENGTH];
		while (reader.getChannel().position() < length)
			result[(int) (reader.getChannel().position() % MAX_LENGTH)] = transformByte((byte)reader.read());
		String accum = "";
		for (int i = 0; i < MAX_LENGTH; i++)
			accum += String.format("%02x", result[i]);
		reader.close();
		return accum;
	}
	
	private byte transformByte(byte bl) {
		a = rotateLeft(bl, x);
		b = (b ^ bl) - x;
		c = (a + b) & x;
		d ^= x - b;
		x ^= d;
		
		return (byte) ((a * c) + b - x * d ^ bl);
	}
	
	private byte rotateLeft(byte b, int bits) {
		return (byte) ((b << bits) | b >> 32 - bits);
	}
	
	private void init() {
		a = 0x6B87;
		b = 0x7F43;
		c = 0xA4Ad;
		d = 0xDC3F;
		x = 0;
	}
	
	private byte[] pad(byte[] bytes) {
		if (bytes.length >= MAX_LENGTH)
			return bytes;
		byte[] result = new byte[MAX_LENGTH];
		for (int i = 0; i < bytes.length; i++)
			result[i] = bytes[i];
		for (int i = bytes.length; i < result.length; i++) {
			result[i] = FILLER_BYTE;
		}
		return result;
	}
}
